### PR TITLE
Issue #1068 average well rate timeseries

### LINKED ISSUE #1068: average well rate timeseries.
Whether to modify `create_time_discretization` or create an extra method which enforces timesteps for wells.  
Might be able to use pandas resampling functionality: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.resample.html.

### PR DESCRIPTION
**Pull request by luitjansl**
Fixes #1068 

# Description
Implements averaging of well rate timeseries onto a set of desired times. The imported well will have rate changes at these desired times. This can be both finer and coarser than the input timeseries. These times do not need to be equidistant. 
With this fix, when importing a project from imod5 data, we first choose the output temporal discretization, and the 
well timeseries on output contain these times only.

# Checklist

- [X] Links to correct issue
- [ ] Update changelog, if changes affect users
- [X] PR title starts with ``Issue #nr``, e.g. ``Issue #737``
- [X] Unit tests were added
- [ ] **If feature added**: Added/extended example

### CODE REVIEW

File: imod/tests/test_mf6/test_mf6_resampling.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+        {"time": datetime(1989, 1, 5), "rate": 400, "x": 0, "y": 0,"id": "ID", "filt_top": 20, "filt_bot": 10, },  # output time 2
+        { "time": datetime(1989, 1, 6), "rate": 500, "x": 0,  "y": 0,"id": "ID","filt_top": 20, "filt_bot": 10,}, # output time 3
+    ]  
+    # fmt: on
+    # initialize data of lists.
+
+    # setup panda table
+    timeseries = pd.DataFrame(
+        data, columns=["time", "rate", "x", "y", "id", "filt_top", "filt_bot"]
+    )
```
Comment (JoerivanEngelen) [inline imod/tests/test_mf6/test_mf6_resampling.py:25]: This is more readable, could you try and see if this works? If so, update the other cases in this form? I think this is more readable.

```suggestion
    timeseries = pd.DataFrame(
        {}, columns=["time", "rate", "x", "y", "id", "filt_top", "filt_bot"]
    )
   
    timeseries["time"] = [datetime(1989, 1, i) for i in [1, 3, 4, 5, 6]]
    timeseries["rate"] = [i * 100 for i in range(1, 6)]
    timeseries["x"] = 0
    timeseries["y"] = 0
    timeseries["id"] = "ID"
    timeseries["filt_top"] = 20
    timeseries["filt_bot"] = 10

```

File: imod/tests/test_mf6/test_mf6_resampling.py
Comment (JoerivanEngelen) [inline imod/tests/test_mf6/test_mf6_resampling.py:1]: Tests look good, nice test cases. My main comments are the test cases setup being quite verbose. I have a suggestion to reduce that, so we can more easily tell what's what. Furthermore the appropriate place for this file is I think is: ``tests/test_mf6/test_utilities/test_resampling.py``

File: imod/util/expand_repetitions.py [hunk 2]
```diff
@@ -15,3 +42,94 @@ def expand_repetitions(
...
+        )
+        intermediate_df.loc[intermediate_df["time"] < well_rate["time"][0], "id"] = (
+            well_rate["id"][0]
+        )
+        intermediate_df.loc[
+            intermediate_df["time"] < well_rate["time"][0], "filt_top"
+        ] = well_rate["filt_top"][0]
+        intermediate_df.loc[
+            intermediate_df["time"] < well_rate["time"][0], "filt_bot"
+        ] = well_rate["filt_bot"][0]
```
Comment (JoerivanEngelen) [inline imod/util/expand_repetitions.py:93]: Getting verbose here, I think this should reduce the amount of code here:

```suggestion
    time_before_start_input = intermediate_df["time"] < well_rate["time"][0]
    if time_before_start_input[0]:
        intermediate_df.loc[time_before_start_input, "rate"] = 0.0
        intermediate_df.loc[time_before_start_input, "x"] = well_rate["x"][0]
        intermediate_df.loc[time_before_start_input, "y"] = well_rate["y"][0]
...
#etc
```

**Review by JoerivanEngelen — COMMENTED** [review]
This my review part 1, stopped at line 94 of expand_repetitions.py; will work on that tomorrow. Submitted now to ensure it is saved and not deleted somewhere.

**Review by JoerivanEngelen — CHANGES_REQUESTED** [review]
I've quite significantly adapted ``imod/util/expand_repetitions.py``, to better utilize pandas functionality to compute the integrals. The unittests for the resampling work now.
Some things that appeared while working on this:

- The tests specify rate as integers, but I think it is more common to have floats. Could you turn them into floats?
- I'm not sure ``resample_timeseries`` are in the right place now as this is namespace is used for quite generic utilities, used throughout different modules. ``Resample_timeseries`` on the other hand is quite specific to resampling a single well. Perhaps a better place for this is: ``imod/mf6/utilities/resampling/wells.py``
- Could you improve the setup of the test cases with what I mentioned?

File: imod/util/expand_repetitions.py [hunk 3]
```diff
@@ -15,3 +42,94 @@ def expand_repetitions(
...
+
+    # compute time difference from perious to current row
+    time_diff_col = intermediate_df["time"].diff()
+    intermediate_df.insert(7, "time_to_next", time_diff_col.values)
```
[the remark this answers is not in the document]
Comment (luitjansl) [reply imod/util/expand_repetitions.py:89]: It's the "insertion index" or the column index where the new column should go. We expect these columns to exist: "time", "rate", "x", "y", "id", "filt_top", "filt_bot". The new column should be added right of "filt _bot" and therefore gets index 7.

File: imod/mf6/model_gwf.py [hunk 4]
```diff
@@ -177,8 +178,7 @@ def from_imod5_data(
         period_data: dict[str, list[datetime]],
         allocation_options: SimulationAllocationOptions,
         distributing_options: SimulationDistributingOptions,
-        time_min: datetime,
-        time_max: datetime,
+        times: list[datetime],
```
[the remark this answers is not in the document]
Comment (luitjansl) [reply imod/mf6/model_gwf.py:182]: according to stack overflow, it's the other way around. Importing from the typing library is deprecated. see here: https://stackoverflow.com/questions/24853923/type-hinting-a-collection-of-a-specified-type/62775680#62775680

File: imod/tests/test_mf6/test_mf6_resampling.py
Comment (luitjansl) [reply imod/tests/test_mf6/test_mf6_resampling.py:1]: done, moved the testfile to tests/test_mf6/test_utilities/test_resampling.py and removed the tables in it.

File: imod/mf6/simulation.py [hunk 5]
```diff
@@ -1325,8 +1326,7 @@ def from_imod5_data(
         period_data: dict[str, dict[str, GridDataArray]],
         allocation_options: SimulationAllocationOptions,
         distributing_options: SimulationDistributingOptions,
-        time_min,
-        time_max,
+        times: list[datetime],
```
[the remark this answers is not in the document]
Comment (luitjansl) [reply imod/mf6/simulation.py:1329]: issue #1111 was created for this

**Review by JoerivanEngelen — APPROVED** [review]
Looks good! Approved!

**Pull request merged**